### PR TITLE
Replacing the color picker with a Hue-Saturation color wheel.

### PR DESCRIPTION
**Pull request by NovapaX**
I've been working on a replacement of the color picker. this PR is the result.

This replaces the color picker with a color wheel. I put a lot of work in making it touch-friendly. It doesn't interfere with scrolling on touch-screens, if you want to drag around the wheel you should 'pick' the marker to start dragging. (on mouse-devices you can click-and-drag anywhere)
It also shows a circle with the 'current' color above your fingers when you're dragging.

Update frequency is set with the 'throttle' attribute. Set it to the same 500ms that was used with the old color picker.
Maybe 200ms is nicer, but I have no idea how several light platforms would react on such a speedy inflow of updated states.

There are some options for styling and color segments. I settled on 24 hue-segments and 8 saturation segments. This way it fits the 'material design' a bit better. But it also restricts the possible colors that can be chosen. (to 7x24+1=169 colors)
If that is undesired we could add the ignore-segments attribute (but that would be a little confusing, because you can select colors that are not even shown) Or we could just completely leave out the segments attributes and show it as a traditional color wheel with smooth gradients.
(styling and segments are done by the 'host' which is including the color picker.)

This picker does Hue and saturation, as the brightness is (or at least should be) controlled with the brightness slider.
Since the backend is still based on RGB, the chosen Hue and Saturation are converted to RGB with a maximum brightness.
Platforms that rely on setting brightness by choosing a 'dark' RGB color should really just use the slider.
There has been a lot of talk about improving the light platform. (and even migrating to separate hue, saturation and brightness attributes). This can be easily converted to return those values.

Please test it out, any comments on both code and functionality are more than welcome.
I'm not a javascript ninja, so forgive me any (noob) mistakes.

schermafbeelding 2017-11-27 om 15 56 46
schermafbeelding 2017-11-27 om 16 23 43

### CODE REVIEW

File: src/components/ha-color-picker.html [hunk 1]
```diff
@@ -5,148 +5,612 @@
...
         cursor: crosshair;
+        width: 100%;
+        height: 100%;
+        overflow: visible;
+      }
+      #backgroundLayer {
+        width: 100%;
+        overflow: visible;
+        --wheel-bordercolor: var(--ha-color-picker-wheel-bordercolor, white);
+        --wheel-borderwidth: var(--ha-color-picker-wheel-borderwith, 3);
```
Comment (balloob) [inline src/components/ha-color-picker.html]: It should be `width` inside `--ha-color-picker-wheel-borderwith`

File: src/components/ha-color-picker.html [hunk 2]
```diff
@@ -5,148 +5,612 @@
...
+        --wheel-shadow: var(--ha-color-picker-wheel-shadow, rgb(15, 15, 15) 10px 5px 5px 0px);
       }
-    </style>
 
-    <canvas width='[[width]]' height='[[height]]' id='canvas'></canvas>
+      #marker {
+        fill: currentColor;
+        stroke: var(--ha-color-picker-marker-bordercolor, white);
+        stroke-width: var(--ha-color-picker-marker-borderwith, 3);
```
Comment (balloob) [inline src/components/ha-color-picker.html]: `width`

File: src/more-infos/more-info-light.html [hunk 3]
```diff
@@ -89,7 +94,13 @@
           on-change='wvSliderChanged'></ha-labeled-slider>
       </div>
 
-      <ha-color-picker on-colorselected='colorPicked' height='200' width='250'>
+      <ha-color-picker
+        on-colorselected='colorPicked'
+        desired-rgb-color='{{colorPickerColor}}'
+        throttle=500
```
Comment (balloob) [inline src/more-infos/more-info-light.html]: Please use quotes around all attributes.

**Comment by balloob** [comment]
This is really cool 🙇

**Comment by balloob** [comment]
- My fingers are a bit too big and the circle that shows above it can end up being hidden by it. My fingers are above average.
 - I would expect to be able to press on an empty space and start dragging and get the same experience as with dragging the badge.

**Comment by NovapaX** [comment]
I can fix the first point by moving that circle a bit higher.
That last point is a bit tough to solve, because that would interfere with scrolling (especially on smaller screen devices) and that is really annoying. (catching touch moves on such a big area in a scrollable container is kind of bad for UX imo.)
It does work the way you describe with a mouse by the way.
I can experiment with some kind of delay where you can hold for a little while (like 200ms, or whatever I end up with) and then it will select and allow dragging. That way it won't interfere with the (mostly faster) scroll gesture.

**Comment by NovapaX** [comment]
I can't get 'touchAnywhere-and-drag' working in a clean way without it glitching on mobile devices and interfering with the tap-and-hold contextual menu popping up or (text)selection triggering.
The best way to get what you want would be to just double tap and drag on the wheel. (first tap selects and puts the marker under your thumb, that way with the second tap you grab the marker.)

I really oppose to making it default to tap-and-drag anywhere because scrolling usually happens at the bottom of the screen, and that's where the wheel appears when you scroll. It's no fun when you just want to scroll down and suddenly your light turns purple. (I've had that too many times with the old color picker, and also with this wheel before I changed the dragging behavior)

**Comment by balloob** [comment]
okidoki

**Comment by balloob** [comment]
🐬 🌮

**Pull request merged**